File: bindings/javascript/src/amqp.js
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>
//
// AMQP protocol types for proven-servers.

/** FrameType matching the Idris2 ABI tags. */
export const FrameType = Object.freeze({
  METHOD: 0,
  HEADER: 1,
  BODY: 2,
  HEARTBEAT: 3,
});

/** MethodClass matching the Idris2 ABI tags. */
export const MethodClass = Object.freeze({
  CONNECTION: 0,
  CHANNEL: 1,
  EXCHANGE: 2,
  QUEUE: 3,
  BASIC: 4,
  TX: 5,
  CONFIRM: 6,
});

/** ExchangeType matching the Idris2 ABI tags. */
export const ExchangeType = Object.freeze({
  DIRECT: 0,
  FANOUT: 1,
  TOPIC: 2,
  HEADERS: 3,
});

/** DeliveryMode matching the Idris2 ABI tags. */
export const DeliveryMode = Object.freeze({
  NON_PERSISTENT: 0,
  PERSISTENT: 1,
});

/** ErrorSeverity matching the Idris2 ABI tags. */
export const ErrorSeverity = Object.freeze({
  CHANNEL_LEVEL: 0,
  CONNECTION_LEVEL: 1,
});

/** ConnectionState matching the Idris2 ABI tags. */
export const ConnectionState = Object.freeze({
  CONNECTION_STATE_IDLE: 0,
  NEGOTIATING: 1,
  TUNING_OK: 2,
  OPEN: 3,
  CLOSING: 4,
});

/** ChannelState matching the Idris2 ABI tags. */
export const ChannelState = Object.freeze({
  CLOSED: 0,
  OPENING: 1,
  CH_OPEN: 2,
  CH_CLOSING: 3,
});

/** BrokerState matching the Idris2 ABI tags. */
export const BrokerState = Object.freeze({
  BROKER_STATE_IDLE: 0,
  CONNECTED: 1,
  CHANNEL_OPEN: 2,
  CONSUMING: 3,
  PUBLISHING: 4,
  DISCONNECTING: 5,
});

File: bindings/dart/lib/src/amqp.dart
// SPDX-License-Identifier: MPL-2.0
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>
//
// AMQP protocol types for proven-servers.

/// FrameType matching the Idris2 ABI tags.
enum FrameType {
  method(0),
  header(1),
  body(2),
  heartbeat(3);

  const FrameType(this.tag);
  final int tag;

  static FrameType? fromTag(int tag) {
    for (final v in values) {
      if (v.tag == tag) return v;
    }
    return null;
  }
}

/// MethodClass matching the Idris2 ABI tags.
enum MethodClass {
  connection(0),
  channel(1),
  exchange(2),
  queue(3),
  basic(4),
  tx(5),
  confirm(6);

  const MethodClass(this.tag);
  final int tag;

  static MethodClass? fromTag(int tag) {
    for (final v in values) {
      if (v.tag == tag) return v;
    }
    return null;
  }
}

/// ExchangeType matching the Idris2 ABI tags.
enum ExchangeType {
  direct(0),
  fanout(1),
  topic(2),
  headers(3);

  const ExchangeType(this.tag);
  final int tag;

  static ExchangeType? fromTag(int tag) {
    for (final v in values) {
      if (v.tag == tag) return v;
    }
    return null;
  }
}

/// DeliveryMode matching the Idris2 ABI tags.
enum DeliveryMode {
  nonPersistent(0),
  persistent(1);

  const DeliveryMode(this.tag);
  final int tag;

  static DeliveryMode? fromTag(int tag) {
    for (final v in values) {
      if (v.tag == tag) return v;
    }
    return null;
  }
}

/// ErrorSeverity matching the Idris2 ABI tags.
enum ErrorSeverity {
  channelLevel(0),
  connectionLevel(1);

  const ErrorSeverity(this.tag);
  final int tag;

  static ErrorSeverity? fromTag(int tag) {
    for (final v in values) {
      if (v.tag == tag) return v;
    }
    return null;
  }
}

/// ConnectionState matching the Idris2 ABI tags.
enum ConnectionState {
  connectionState_Idle(0),
  negotiating(1),
  tuningOk(2),
  open(3),
  closing(4);

  const ConnectionState(this.tag);
  final int tag;

  static ConnectionState? fromTag(int tag) {
    for (final v in values) {
      if (v.tag == tag) return v;
    }
    return null;
  }
}

/// ChannelState matching the Idris2 ABI tags.
enum ChannelState {
  closed(0),
  opening(1),
  chOpen(2),
  chClosing(3);

  const ChannelState(this.tag);
  final int tag;

  static ChannelState? fromTag(int tag) {
    for (final v in values) {
      if (v.tag == tag) return v;
    }
    return null;
  }
}

/// BrokerState matching the Idris2 ABI tags.
enum BrokerState {
  brokerState_Idle(0),
  connected(1),
  channelOpen(2),
  consuming(3),
  publishing(4),
  disconnecting(5);

  const BrokerState(this.tag);
  final int tag;

  static BrokerState? fromTag(int tag) {
    for (final v in values) {
      if (v.tag == tag) return v;
    }
    return null;
  }
}
